### DISCUSSION TITLE
Unable to get the tatweel flag information for the glyphs from hb-shape tool (HarfBuzz 7.3.0 & 8.3.0)

### DISCUSSION THREAD

**Discussion by sid-wr** [harfbuzz/harfbuzz#4528] [General] [answered]
Hi @behdad @khaledhosny 

I tried the command line hb-shape.exe tool with multiple fonts for Arabic input as follows:

```
hb-shape.exe --safe-to-insert-tatweel NotoNaskhArabic-VariableFont_wght.ttf -u "U+0627U+0644U+0623U+0632DU+0631U+0641"
hb-shape.exe --safe-to-insert-tatweel NotoSansArabic-VariableFont_wdth,wght.ttf -u "U+0627U+0644U+0623U+0632DU+0631U+0641"
```

But I am unable to see any such extra information per glyph.

I also tried hb-shape.exe without --safe-to-insert-tatweel flag as follows , & the results are same as that with that flag
```
hb-shape.exe NotoNaskhArabic-VariableFont_wght.ttf -u "U+0627U+0644U+0623U+0632DU+0631U+0641"
hb-shape.exe NotoSansArabic-VariableFont_wdth,wght.ttf -u "U+0627U+0644U+0623U+0632DU+0631U+0641"
```

This is the output for both these sets of commands (Does not seem to be any tatweel related information here as per me):
```
NotoNaskhArabic -> [uni0641=5+848|uni0631=4+386|.notdef=3+646|uniFE84.rlig=2+0|uni0644.rlig=1+518|uni0627=0+238]
NotoSansArabic -> [uni0641=5+953|uni0631=4+367|.notdef=3+600|uniFEF7=1+582|uni0627=0+238]
```

I hope I am not missing any required step or command line argument.

Please help
Thanks

**Comment by khaledhosny** [comment, marked as answer]
You still need `--show-flags` to show any flags (`--safe-to-insert-tatweel` tells HarfBuzz to calculate this flag while shaping).

**Reply by khaledhosny** [reply]
(Also, there is a typo in your input, `U+0632D` has an extra `D` at the end.)

**Comment by sid-wr** [comment]
Thanks for pointing out the typo & the other flag, I am able to see that info by using `--show-flags` and `--safe-to-insert-tatweel` together

**Comment by sid-wr** [comment]
However, I can see some inconsistencies with different fonts.

For example:
```
hb-shape.exe --show-flags --safe-to-insert-tatweel GeezaPro.ttc -u “U+0627U+0644U+0623U+062DU+0631U+0641”
hb-shape.exe --show-flags --safe-to-insert-tatweel NotoSansArabic-VariableFont_wdth,wght.ttf -u “U+0627U+0644U+0623U+062DU+0631U+0641"
```

This gives the following output:
```
GeezaPro           -> [u0641.feh=5+1733#1|u0631.final.reh=4@-202,0+700#1|u062d.initial.hah=3+1359#1|u0644_u0623.isolated.lamHamzaOnAlef=1+1162|u0627.alef=0+647]
Noto Sans Arabic   -> [uni0641=5+953#1|uniFEAE=4+396#5|uniFEA3=3+589|uniFEF7=1+582|uni0627=0+238]
```

As per my understanding, the tatweel flag [defined in the code](https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-buffer.h#L163C3-L163C39) is getting denoted by the suffixes `#4` or `#5` or `#7` computed after bit-wise AND/OR.

The above output indicates that the tatweel flag  is safe to add before 2nd last glyph for Noto font, but not in case of GeezaPro font

But if we visually observe the glyphs as follows, then Tatweel should be allowed for both the fonts:
![Doubt](https://github.com/harfbuzz/harfbuzz/assets/136314541/95105d84-08c1-4e38-a71d-4ef803821139)

Please correct me if I am wrong

**Reply by behdad** [reply]
GeezaPro is an Apple AAT `morx` font. We don't implement tatweel for it. And it's almost impossible to do.

**Comment by sid-wr** [comment]
Hi @khaledhosny 
I have marked your comment as answer, but can you please have a look at this inconsistency
Thanks
